### ISSUE TITLE
Change sigchain claim IDs to be timestamped "flake IDs"

### ISSUE THREAD

**Issue by joshuakarp** [MatrixAI/Polykey#195]
We currently use https://github.com/substack/lexicographic-integer to allow the claims in the sigchain to be numerically ordered whilst in LevelDB (that uses a lexicographical ordering on their keys). This allows us to easily iterate across adjacent entries in the sigchain.

However, if we eventually want to replicate our sigchain across different nodes in a gestalt (or across Polykey nodes in general), we need to introduce some uniqueness to these sequence numbers.

The `monotonic-lexicographic-timestamp` library would allow us to do so https://github.com/vweevers/monotonic-lexicographic-timestamp: however, we would need the following issue to be resolved: https://github.com/vweevers/monotonic-lexicographic-timestamp/issues/3

**Comment by CMCDragonkai** [comment]
This should be solvable with the new https://github.com/MatrixAI/js-id being released.

**Comment by CMCDragonkai** [comment]
Going to be addressed in: https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205#note_700077232

**Comment by joshuakarp** [comment]
It's worthwhile to note the distinction between the claim ID, and the sequence number.

The sequence number forms part of the "claim" itself that's stored in the sigchain. It refers to the position of the claim in a node's sigchain (the first claim has a sequence number of 1). That is, it's part of:

```typescript
type Claim = {
  payload: {
    hPrev: string | null; // Hash of the previous claim (null if first claim)
    seq: number; // Sequence number of the claim
    data: ClaimData; // Our custom payload data
    iat: number; // Timestamp (initialised at JWS field)
  };
  signatures: Record<NodeId, SignatureData>; // Signee node ID -> claim signature
};
```

The `ClaimId` is the unique identifier for the claim (used as the key in the sigchain's database). Currently, we're using a lexi-int to make these sortable (simply packing the sequence number as a lexi-int): 

```typescript
/**
 * An arbitraty string serving as a unique identitifer for a particular claim.
 * Depending on the domain the claim is used in, its implementation detail will
 * differ. For example, the sigchain domain uses a lexicographic-integer as the
 * claim ID (representing the sequence number key of the claim).
 */
type ClaimId = Opaque<'ClaimId', string>;
```

So, in summary, the sequence number used to be the "seed" for the claim ID. I don't believe the `IdSortable` from Roger's library will take a seed like this, but it's worthwhile to know.

**Comment by CMCDragonkai** [comment]
So:

1. The `seq` is currently the same value used for `ClaimId`, this is not a hard requirements. The `ClaimId` can now be using `IdSortable`, and the `seq` can stay as a sequence counter representing both ordinal and cardinal properties of the sigchain.
2. The `iat` is a timestamp. It is possible to extract the timestamp out of any `IdSortable` with `extractTs`. This will give you back something like `123.124` where `123` is seconds and `124` is milliseconds. So if `iat` is in seconds form, just truncate it. If it in millisecond form, just multiply the ts by `1000` and you'll have the millisecond timestamp.

**Comment by CMCDragonkai** [comment]
The `iat` in seconds, so you just need to truncate the extracted timestamp with `Math.trunc`.

**Comment by tegefaulkes** [comment]
I think this issue is completed due to the ClaimId being a `IdSortable` from `@matrixai/id` now.

**Comment by joshuakarp** [comment]
Yeah, closing this now - changes from !205 have resolved this https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205/diffs?commit_id=8fc4003b8d58a636210a1ad83d0da02531410101.
